hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientReliableTopicProxy.java

```diff
@@ -118,6 +129,15 @@ public void publish(@Nonnull E payload) {
         }
     }
 
+    @Override
+    public InternalCompletableFuture<E> publishAsync(@Nonnull E message) {
+        checkNotNull(message, NULL_MESSAGE_IS_NOT_ALLOWED);
```

Comment:
Hello @mmedenjak, sorry, but I fail with the `ClientReliableTopicProxy#addWithBackoff` Method. This is a synchronous method and I actually have no idea how to implement an asynchronous one. Do you have an idea?

Reply:
That one is problematic, true. I guess `BLOCK` is not really a good name for the policy but maybe you can create a future and a self-scheduling task which will reschedule itself if it cannot add the item. Once it manages to add the item, it completes the task. So calling `get()` on such a future would yield the same behaviour as the sync variant. Not sure if it makes sense and let me know if you need help with an example of how I see it.

Reply:
Hello @mmedenjak, I know what you mean, but do you have an example of such a self-scheduling task?

Reply:
Hey, I was off for three weeks, I'll try to figure out an example and send it to you.

Reply:
Hey Matko! I can give you a little bit more information about the generics problem.. 
The problem is, that every RingBuffer method returns a `CompletableFuture<Long>` and not a `CompletableFuture<E>`.

``` java
    @Override
    public CompletableFuture<E> publishAllAsync(@Nonnull Collection<? extends E> payloads) {
        checkNotNull(payloads, NULL_MESSAGE_IS_NOT_ALLOWED);
        checkNoNullInside(payloads, NULL_MESSAGE_IS_NOT_ALLOWED);
        CompletableFuture<Long> completableFuture = null;
        try {
            Collection<ReliableTopicMessage> messages = new ArrayList<>(payloads.size());
            payloads.forEach(payload -> messages.add(new ReliableTopicMessage(toData(payload), null)));
            switch (overloadPolicy) {
                case ERROR:
                case DISCARD_NEWEST:
                    completableFuture = ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL).toCompletableFuture();
                    break;
                case DISCARD_OLDEST:
                    completableFuture = ringbuffer.addAllAsync(messages, OverflowPolicy.OVERWRITE).toCompletableFuture();
                    break;
                case BLOCK:
                    completableFuture = addWithBackoffAsync(messages);
                    break;
                default:
                    throw new IllegalArgumentException("Unknown overloadPolicy:" + overloadPolicy);
            }
        } catch (Exception e) {
            throw (RuntimeException) peel(e, null,
                    "Failed to publish message: " + payloads + " to topic:" + getName());
        }
        return new completableFuture;
    }
```

Reply:
Hey, not sure if you might still be willing to work on this issue but I managed to take a look and here are my conclusions. I believe we should change all return types for `publishAsync` and `publishAllAsync` to these:
```
CompletionStage<Void> publishAsync(@Nonnull E message);
CompletionStage<Void> publishAllAsync(@Nonnull Collection<? extends E> messages);
```

The reasons are:
- all other async methods on our public API return `CompletionStage` and none return `InternalCompletableFuture` since it's an internal implementation
- the generic type should probably be `Void` since different implementations of `ITopic` use different backing structures and since `ITopic#publish` returns `void` anyway and not the ID or the message. Of course, we could return the messages themselves and I'm sure each implementation may create a new future which will be completed with the provided messages but I don't think that's necessary. The messages were provided as a parameter to the async methods which means the user already has references to them. What matters is the future itself and being able to react to the completion and not the messages themselves.

This also means you will have to adapt some methods to return a newly created future which will be completed when some operation completes. For instance, in `ReliableTopicProxy#publishAsync`:

```
    @Override
    public CompletionStage<Void> publishAsync(@Nonnull E message) {
        checkNotNull(message, NULL_MESSAGE_IS_NOT_ALLOWED);
        Data data = toData(message);
        InternalCompletableFuture<Void> returnFuture = new InternalCompletableFuture<>();
        invokeOnPartition(new PublishOperation(name, data))
                .whenCompleteAsync((o, t) -> {
                    if (t != null) {
                        returnFuture.completeExceptionally(t);
                    } else {
                        returnFuture.complete(null);
                    }
                });
        return returnFuture;
    }
```

Note: I haven't tested this but I believe it works correctly, you'll need to verify this.

Reply:
As for the self-scheduling and other details of the `ReliableTopicProxy` implementation, here's what I came up with:
```
@Override
    public void publishAll(@Nonnull Collection<? extends E> payload) {
        checkNotNull(payload, NULL_MESSAGES_IS_NOT_ALLOWED);
        checkNoNullInside(payload, NULL_MESSAGE_IS_NOT_ALLOWED);

        try {
            List<ReliableTopicMessage> messages = payload.stream()
                                                         .map(m -> new ReliableTopicMessage(nodeEngine.toData(m), thisAddress))
                                                         .collect(Collectors.toList());
            switch (overloadPolicy) {
                case ERROR:
                    long sequenceId = ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL).toCompletableFuture().get();
                    if (sequenceId == -1) {
                        throw new TopicOverloadException("Failed to publish messages: " + payload + " on topic:" + getName());
                    }
                    break;
                case DISCARD_OLDEST:
                    ringbuffer.addAllAsync(messages, OverflowPolicy.OVERWRITE).toCompletableFuture().get();
                    break;
                case DISCARD_NEWEST:
                    ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL).toCompletableFuture().get();
                    break;
                case BLOCK:
                    long timeoutMs = INITIAL_BACKOFF_MS;
                    for (; ; ) {
                        long result = ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL).toCompletableFuture().get();
                        if (result != -1) {
                            break;
                        }

                        MILLISECONDS.sleep(timeoutMs);
                        timeoutMs *= 2;
                        if (timeoutMs > MAX_BACKOFF) {
                            timeoutMs = MAX_BACKOFF;
                        }
                    }
                    break;
                default:
                    throw new IllegalArgumentException("Unknown overloadPolicy:" + overloadPolicy);
            }

            localTopicStats.incrementPublishes();
        } catch (Exception e) {
            throw (RuntimeException) peel(e, null, "Failed to publish messages: " + payload + " to topic:" + getName());
        }
    }


@Override
    public CompletionStage<Void> publishAllAsync(@Nonnull Collection<? extends E> payload) {
        checkNotNull(payload, NULL_MESSAGES_IS_NOT_ALLOWED);
        checkNoNullInside(payload, NULL_MESSAGE_IS_NOT_ALLOWED);
        InternalCompletableFuture<Void> returnFuture = new InternalCompletableFuture<>();

        try {
            List<ReliableTopicMessage> messages = payload.stream()
                                                         .map(m -> new ReliableTopicMessage(nodeEngine.toData(m), thisAddress))
                                                         .collect(Collectors.toList());
            switch (overloadPolicy) {
                case ERROR:
                    ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL)
                              .whenCompleteAsync((id, t) -> {
                                  if (t != null) {
                                      returnFuture.completeExceptionally(t);
                                  }
                                  if (id == -1) {
                                      returnFuture.completeExceptionally(new TopicOverloadException(
                                              "Failed to publish messages: " + payload + " on topic:" + getName()));
                                  }
                                  localTopicStats.incrementPublishes(payload.size());
                                  returnFuture.complete(null);
                              });

                    break;
                case DISCARD_OLDEST:
                    ringbuffer.addAllAsync(messages, OverflowPolicy.OVERWRITE)
                              .whenCompleteAsync((id, t) -> {
                                  if (t != null) {
                                      returnFuture.completeExceptionally(t);
                                  }
                                  localTopicStats.incrementPublishes(payload.size());
                                  returnFuture.complete(null);
                              });
                    break;
                case DISCARD_NEWEST:
                    ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL)
                              .whenCompleteAsync((id, t) -> {
                                  if (t != null) {
                                      returnFuture.completeExceptionally(t);
                                  }
                                  localTopicStats.incrementPublishes(payload.size());
                                  returnFuture.complete(null);
                              });
                    break;
                case BLOCK:
                    addAsyncAndBlock(payload, returnFuture, messages, INITIAL_BACKOFF_MS);
                    break;
                default:
                    throw new IllegalArgumentException("Unknown overloadPolicy:" + overloadPolicy);
            }
        } catch (Exception e) {
            returnFuture.completeExceptionally(peel(e, null, "Failed to publish messages: " + payload + " to topic:" + getName()));
        }

        return returnFuture;
    }

    private void addAsyncAndBlock(@Nonnull Collection<? extends E> payload,
                                  InternalCompletableFuture<Void> returnFuture,
                                  List<ReliableTopicMessage> messages,
                                  long pauseMillis) {
        ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL).whenCompleteAsync((id, t) -> {
            if (t != null) {
                returnFuture.completeExceptionally(t);
            }
            if (id == -1) {
                nodeEngine.getExecutionService().schedule(
                        () -> addAsyncAndBlock(payload, returnFuture, messages, Math.min(pauseMillis * 2, MAX_BACKOFF)),
                        pauseMillis, MILLISECONDS);
                returnFuture.completeExceptionally(new TopicOverloadException(
                        "Failed to publish messages: " + payload + " on topic:" + getName()));
            }
            localTopicStats.incrementPublishes(payload.size());
            returnFuture.complete(null);
        });
    }
```

I'm sure it can be cleaned up a bit and it definitely needs to be tested but that's what I was thinking about.